src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/Snapshots/ImmutablePropertyCollection.cs

```diff
@@ -31,21 +32,23 @@ protected ImmutablePropertyCollection(ImmutableArray<U> inputItems, Func<U, stri
             // If the input is empty, don't allocate anything further.
             if (inputItems.Length == 0)
             {
-                _items = Array.Empty<T>();
+                _itemByIndex = Array.Empty<T>();
                 _itemsByName = ImmutableDictionary<string, T>.Empty;
                 return;
             }
 
-            // Build a list, to maintain order for index-based lookup.
-            var itemList = new List<T>();
-            
+            // Build an array, to maintain order for index-based lookup.
+            var itemByIndex = new T[inputItems.Length];
+
             // Build a dictionary to support key-based lookup.
-            var itemByName = new Dictionary<string, T>();
-            
+            var itemByName = new Dictionary<string, T>(capacity: inputItems.Length);
+
+            int index = 0;
+
             foreach (U inputItem in inputItems)
```

Comment:
Use for loop? We need to maintain indexer already and ImmutableArray has accessor.

Reply:
Interesting. We do foreach over `ImmutableArray<>` in a lot of places as the code's clearer.

Here's a benchmark:

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22621
Intel Xeon Silver 4110 CPU 2.10GHz, 2 CPU, 32 logical and 16 physical cores
  [Host]             : .NET Framework 4.8 (4.8.9139.0), X64 RyuJIT
  .NET Framework 4.8 : .NET Framework 4.8 (4.8.9139.0), X64 RyuJIT

Job=.NET Framework 4.8  Runtime=.NET Framework 4.8
```

|                 Method | Count |          Mean |      Error |     StdDev | Rank | Allocated |
|----------------------- |------ |--------------:|-----------:|-----------:|-----:|----------:|
|     ImmutableArray_For |     0 |     0.7882 ns |  0.0618 ns |  0.0925 ns |    1 |         - |
| ImmutableArray_Foreach |     0 |     4.9222 ns |  0.1027 ns |  0.0911 ns |    3 |         - |
|     ImmutableArray_For |     1 |     1.5354 ns |  0.0781 ns |  0.0869 ns |    2 |         - |
| ImmutableArray_Foreach |     1 |     6.6947 ns |  0.0429 ns |  0.0335 ns |    4 |         - |
|     ImmutableArray_For |    10 |    12.4208 ns |  0.1727 ns |  0.1615 ns |    5 |         - |
| ImmutableArray_Foreach |    10 |    27.0716 ns |  0.2109 ns |  0.1761 ns |    6 |         - |
|     ImmutableArray_For |   100 |   134.3119 ns |  2.0973 ns |  1.6374 ns |    7 |         - |
| ImmutableArray_Foreach |   100 |   234.7997 ns |  3.2206 ns |  2.8550 ns |    8 |         - |
|     ImmutableArray_For |  1000 | 1,184.0567 ns | 17.3725 ns | 15.4003 ns |    9 |         - |
| ImmutableArray_Foreach |  1000 | 2,292.9066 ns |  9.0226 ns |  7.5342 ns |   10 |         - |

So indeed `for` is faster than `foreach`, though the overhead is tiny.

```c#
[SimpleJob(RuntimeMoniker.Net48)]
[RankColumn]
[MemoryDiagnoser]
public class ImmutableArrayIterationBenchmark
{
    private ImmutableArray<string> strings;

    [Params(0, 1, 10, 100, 1000)]
    public int Count;

    [GlobalSetup]
    public void Setup()
    {
        strings = Enumerable.Repeat("Foo", Count).ToImmutableArray();
    }

    [Benchmark]
    public int ImmutableArray_For()
    {
        int count = 0;

        for (int i = 0; i < Count; i++)
        {
            count += strings[i].Length;
        }

        return count;
    }

    [Benchmark]
    public int ImmutableArray_Foreach()
    {
        int count = 0;

        foreach (string s in strings)
        {
            count += s.Length;
        }

        return count;
    }
}
```